File: Java/lab17_Mediator/task_2_1/src/Customer.java
public class Customer {

  private final String address;
  private final ECommerceSite site;

  public Customer(String address) {
    this.address = address;
    site = new ECommerceSite(this);
  }

  public String getAddress() {
    return address;
  }

  public void buy(String item, int quantity) {
    if(site.checkInStock(item, quantity)) {
      site.sell(item, quantity);
    }
  }

}

File: Cpp/lab17_Mediator/task_2_1/src/customer.cpp
#include "customer.hpp"
#include "ecommerce_site.hpp"


Customer::Customer(std::string const& address)
: address(address)
{
  this->site = new ECommerceSite(*this);
}

Customer::Customer(Customer const& customer)
: address(customer.address), site(site) {}

Customer::~Customer()
{
  delete this->site;
}

std::string const& Customer::getAddress() const
{
  return this->address;
}

void Customer::buy(std::string const& item, int quantity)
{
  if (this->site->checkInStock(item, quantity))
  {
    this->site->sell(item, quantity);
  }
}
